### ISSUE TITLE
t/firebird.t fails

### ISSUE THREAD

**Issue by eserte** [sqitchers/sqitch#597]
If DBD::Firebird is installed then t/firebird.t seems to fail (seen on centos8, fedora and debian11 systems, various perl versions):
```
    #   Failed test 'Database should not yet be initialized'
    #   at t/lib/DBIEngineTest.pm line 103.
    # Looks like you failed 1 test of 2.

#   Failed test 'live database'
#   at t/lib/DBIEngineTest.pm line 1780.
Sqitch database /var/tmp/cpansmoker-1023/2021112112/__metasqitch already initialized

Trace begun at /home/cpansand/.cpan/build/2021112112/App-Sqitch-v1.2.0-2/blib/lib/App/Sqitch/Engine/firebird.pm line 206
App::Sqitch::Engine::firebird::initialize('App::Sqitch::Engine::firebird=HASH(0x53b1ee0)') called at t/lib/DBIEngineTest.pm line 108
DBIEngineTest::__ANON__ at /opt/perl-5.28.2t/lib/site_perl/5.28.2/Test/Builder.pm line 374
eval {...} at /opt/perl-5.28.2t/lib/site_perl/5.28.2/Test/Builder.pm line 374
Test::Builder::subtest('Test::Builder=HASH(0x20aa6c0)', 'live database', 'CODE(0x535ae58)') called at /opt/perl-5.28.2t/lib/site_perl/5.28.2/Test/More.pm line 809
Test::More::subtest('live database', 'CODE(0x535ae58)') called at t/lib/DBIEngineTest.pm line 1780
DBIEngineTest::run('DBIEngineTest', 'class', 'App::Sqitch::Engine::firebird', 'target_params', 'ARRAY(0x4825a40)', 'alt_target_params', 'ARRAY(0x482d6b8)', 'skip_unless', 'CODE(0x482d0e8)', 'engine_err_regex', 'Regexp=REGEXP(0x5373018)', 'init_error', 'Sqitch database /var/tmp/cpansmoker-1023/2021112112/__sqitchtest already initialized', 'add_second_format', 'dateadd(1 second to %s)', 'test_dbh', 'CODE(0x4517c78)') called at t/firebird.t line 381
Unsuccessful execution caused by a system error that precludes successful execution of subsequent statements
-I/O error during "open" operation for file "/var/tmp/cpansmoker-1023/2021112112/__sqitchtest"
-Error while trying to open file
-No such file or directory at t/firebird.t line 323.
END failed--call queue aborted.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 82.
t/firebird.t ........ 
Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 1/82 subtests 
```

**Comment by eserte** [comment]
Same test file, but different test output:
```
Statement failed, SQLSTATE = 08001
I/O error during "lock" operation for file "/var/tmp/cpansmoker-1023/2021112112/__metasqitch"
-Database already opened with engine instance, incompatible with current
After line 0 in file /home/cpansand/.cpan/build/2021112112/App-Sqitch-v1.2.0-0/blib/lib/App/Sqitch/Engine/Upgrade/firebird-1.0.sql
"isql-fb" unexpectedly returned exit value 1
Unsuccessful execution caused by a system error that precludes successful execution of subsequent statements
-I/O error during "open" operation for file "/var/tmp/cpansmoker-1023/2021112112/__sqitchtest"
-Error while trying to open file
-No such file or directory at t/firebird.t line 323.
END failed--call queue aborted.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 82.
t/firebird.t ........ 
Dubious, test returned 255 (wstat 65280, 0xff00)
All 82 subtests passed 
```

**Comment by theory** [comment]
Your smoke testing provides an amazing service, @eserte, thanks so much for running it!

Are you running tests in parallel on the same box? Looks like it's connecting to the database, but because both tests use the same database objects, it fails. I noticed the same thing with Snowflake yesterday. I can have it use a different database file every time for the next release. Should probably do that for all of the engine tests.

**Comment by eserte** [comment]
Normally the distributions are not tested in parallel (however, it may happen that I manually run tests on these boxes and then possibly run in parallel with the automatic process, but this was not the case here, I think).

If it matters, I see some __metasqitch leftovers here --- maybe cleanup does not work correctly?

Having a random name for the database file is probably a good idea.

**Comment by theory** [comment]
Fixed in bdc2b0c I believe, thank you!

**Comment by theory** [comment]
Holy crap there were a lot of these!

* [2021-11-21T11:42:38Z](http://www.cpantesters.org/cpan/report/20bc9936-4ac0-11ec-85ce-82381f24ea8f)
* [2021-11-21T12:20:14Z](http://www.cpantesters.org/cpan/report/6156cb4c-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:16Z](http://www.cpantesters.org/cpan/report/62c7857a-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:19Z](http://www.cpantesters.org/cpan/report/642e0e34-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:21Z](http://www.cpantesters.org/cpan/report/659435be-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:23Z](http://www.cpantesters.org/cpan/report/66fbfac2-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:26Z](http://www.cpantesters.org/cpan/report/68674114-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:28Z](http://www.cpantesters.org/cpan/report/69cd9abc-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:30Z](http://www.cpantesters.org/cpan/report/6b3dc160-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:35Z](http://www.cpantesters.org/cpan/report/6e091c5a-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:38Z](http://www.cpantesters.org/cpan/report/6f6fa316-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:40Z](http://www.cpantesters.org/cpan/report/7100e99c-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:20:47Z](http://www.cpantesters.org/cpan/report/753957a6-4ac5-11ec-986e-ae521f24ea8f)
* [2021-11-21T12:21:01Z](http://www.cpantesters.org/cpan/report/7daba9ac-4ac5-11ec-986e-ae521f24ea8f)

The time proximity definitely makes me think that they were stomping on each other. @eserte can you re-run the tests for a couple of these configurations to see if they pass when they're not competing with themselves? Because if not then I'll have to guess at some other solution.

**Comment by eserte** [comment]
It's quite likely that the reports came from different VMs. You can check the hostname part in the uname value in the perl -V output of the report if you want to be sure.

And I am not sure whether the timestamps show the time of report creation or report sending — I think it's the latter. And there's usually some lag between these two times on my systems.

**Comment by theory** [comment]
Pretty weird. Will be curious to see if the issue appears on v1.2.1, too. :-(

**Comment by theory** [comment]
Would I be right that the failures for 1.2.1 are variations o the same issue, @eserte?

Can you tell me how these boxes were configured (OS, packaging, etc.)? I'd like to try to replicate them with Docker images. Thanks.

**Comment by eserte** [comment]
Here's a Dockerfile which can be used to reproduce the problem:
```
# docker build -t perl-test . && docker run perl-test
FROM debian:11
RUN apt-get -y update
RUN apt-get -y --no-install-recommends install perl make cpanminus
RUN apt-get -y --no-install-recommends install libdbd-firebird-perl
RUN apt-get -y --no-install-recommends install libclone-perl libdatetime-perl libdatetime-timezone-perl libio-pager-perl liblist-moreutils-perl libnamespace-autoclean-perl libperlio-utf8-strict-perl libstring-formatter-perl libsub-exporter-perl libtest-mockmodule-perl libtemplate-tiny-perl
RUN apt-get -y --no-install-recommends install libcapture-tiny-perl libconfig-gitlike-perl libencode-locale-perl libhash-merge-perl libintl-perl libipc-run3-perl libipc-system-simple-perl libmoo-perl libpath-class-perl libtest-deep-perl libtest-dir-perl libtest-exception-perl libtest-file-contents-perl libtest-file-perl libtest-mockobject-perl libtest-nowarnings-perl libtest-warn-perl libthrowable-perl libtype-tiny-perl liburi-db-perl liburi-perl
RUN apt-get -y --no-install-recommends install firebird3.0-server-core
CMD cpanm --test-only --verbose DWHEELER/App-Sqitch-v1.2.1.tar.gz
```
(And sorry for not replying earlier...)

**Comment by theory** [comment]
Thank you @eserte. I was able to replicate the issue and fixed it in fc9a35fb. Please give it a try.

That allows the database creation to work, and then the live tests are skipped because `isql` isn't installed. Sadly, installing it leads to a new failure:

```
# Subtest: live database
# Detected ISQL Version: LI-V3.0.7.33374 Firebird 3.0
# Connected to Firebird 3.0.7
    ok 1 - Engine instantiated
    ok 2 - Database should not yet be initialized
    ok 3 - Initialize the database
    ok 4 - Database should now be initialized
    ok 5 - Registry should not need upgrading
    ok 6 - The release should be registered
    ok 7 - Registry should need upgrading
Statement failed, SQLSTATE = 08001
I/O error during "lock" operation for file "/tmp/__sqitchreg_1145o79p"
-Database already opened with engine instance, incompatible with current
After line 0 in file /repo/lib/App/Sqitch/Engine/Upgrade/firebird-1.0.sql
    1..7
ok 82 - live database
"isql-fb" unexpectedly returned exit value 1
Unsuccessful execution caused by a system error that precludes successful execution of subsequent statements
-I/O error during "open" operation for file "/tmp/__metasqitch_1145o79p"
-Error while trying to open file
-No such file or directory at t/firebird.t line 329.
END failed--call queue aborted.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 82.
```

This appears to be because there is no Firebird server running, and there can be [only one file-based connection](https://stackoverflow.com/a/49073660/79202) at a time. So if you were to install the `firebird3.0-utils` package it would start failing again. I am not sure how to fix that.

**Comment by theory** [comment]
Fixed the double connection issue and squashed in af026c11. LMK if you still have failures, @eserte. Note you'll need to install `firebird3.0-utils` to actually run the live Firebird tests.

**Comment by theory** [comment]
@eserte Can you confirm whether af026c1 fixes the issue for you? TIA

**Comment by eserte** [comment]
On the debian 11 system with https://github.com/sqitchers/sqitch/commit/af026c11c3888e875f2a98eb1a99ba06f47fc2d9 checked out:
If firebird3.0-utils is installed then t/firebird.t  passes.
If it is not then it fails with
```
...
ok 77 - DateTime TZ should be set
ok 78 - App::Sqitch::Engine::firebird->can(...)
# Subtest: live database
    1..0 # SKIP Unable to live-test Firebird engine: Unable to locate Firebird ISQL; set "engine.firebird.client" via sqitch config
ok 79 # skip Unable to live-test Firebird engine: Unable to locate Firebird ISQL; set "engine.firebird.client" via sqitch config
1..79
Unsuccessful execution caused by a system error that precludes successful execution of subsequent statements
-I/O error during "open" operation for file "/tmp/__sqitchreg_9m6nzlaj"
-Error while trying to open file
-No such file or directory at t/firebird.t line 329.
END failed--call queue aborted.
# Looks like your test exited with 255 just after 79.
```

**Comment by theory** [comment]
Can't believe I missed that. Fixed in a6bc5ecc.

**Comment by theory** [comment]
Would you mind confirming that a6bc5ec fixes the issue for you, too, @eserte? Many thanks!

**Comment by theory** [comment]
Merged a6bc5ecc.

**Comment by theory** [comment]
Merged in 5f483db7.
